src/file_pipe_log/log_file.rs

```diff
@@ -6,35 +6,114 @@ use std::io::{Read, Seek, SeekFrom, Write};
-use crate::pipe_log::FileBlockHandle;
+use crate::pipe_log::{FileBlockHandle, FileSeq};
 use crate::{Error, Result};
 
-use super::format::LogFileHeader;
+use super::format::{LogFileHeader, Version};
 use crate::env::{FileSystem, Handle, WriteExt};
 
 /// Maximum number of bytes to allocate ahead.
 const FILE_ALLOCATE_SIZE: usize = 2 * 1024 * 1024;
 
+/// Combination of `[Handle]` and `[Version]`, specifying a handler of a file.
+#[derive(Clone, Debug)]
+pub struct FileHandler<F: FileSystem> {
+    pub seq: FileSeq,
+    pub handle: Arc<F::Handle>,
+    pub version: Option<Version>,
+}
+
+/// Function for reading the header of the log file, and return a
+/// `[LogFileHeader]`.
+///
+/// Attention please, to avoid to move the offset of the given `[handle]`, we
+/// use a copy of the `[handle]` to parse the header of the given file.
+pub(super) fn build_file_header<F: FileSystem>(
+    system: &F,
+    handle: Arc<F::Handle>,
+    expected_reader: Option<&mut F::Reader>,
+) -> Result<LogFileHeader> {
+    let file_size: usize = match handle.file_size() {
+        Ok(size) => size,
+        Err(_) => {
+            return Err(Error::Corruption("Corrupted file!".to_owned())); // invalid file
+        }
+    };
+    // [1] If the file was a new file, we just return the default `LogFileHeader`.
+    if file_size == 0 {
+        return Ok(LogFileHeader::default());
+    }
+    // [2] If the length lessed than the standard `LogFileHeader::len()`.
+    let header_len = LogFileHeader::len();
+    if file_size < header_len {
+        return Err(Error::Corruption("Invalid header of LogFile!".to_owned()));
+    }
+    // [3] Parse the header of the file.
+    let parse_file_header = |handle: Arc<F::Handle>, header_len: usize| -> Result<LogFileHeader> {
+        let mut local_reader;
+        let reader = if let Some(rd) = expected_reader {
+            rd
+        } else {
+            local_reader = system.new_reader(handle)?;
+            &mut local_reader
+        };
+        reader.seek(SeekFrom::Start(0))?; // move to head of the file.
+
+        // Read and parse the header.
+        let mut container = vec![0; header_len as usize];
+        let mut buf = &mut container[..];
+        loop {
+            match reader.read(buf) {
+                Ok(0) => {
+                    break;
+                }
+                Ok(n) => {
+                    buf = &mut buf[n..];
+                }
+                Err(e) => return Err(Error::Io(e)),
+            }
+        }
+        LogFileHeader::decode(&mut container.as_slice())
+    };
+    parse_file_header(handle, header_len)
+}
+
+/// Build a file writer.
+///
+/// * `[handle]`: standard handle of a log file.
+/// * `[rewrite_flag]`:
+///     - `true` => rewrite the header with `[expected_hdr]`;
+///     - `false`=> use the original header from the log file.
 pub(super) fn build_file_writer<F: FileSystem>(
     system: &F,
     handle: Arc<F::Handle>,
+    expected_hdr: Option<LogFileHeader>,
```

Comment:
Two things here,
(1) use version like reader, in the future we might have more options in `LogFileHeader` that could affect file writing, but right now a version is good enough.
(2) Always provide a valid version to build file writer. It's because extracting version requires creating a file reader. It is sort-of a heavy operation, but we can't infer that information from the function name `build_file_writer`.
```suggestion
    version: Version,
```